### PR TITLE
Fixing missing commas on serialization

### PR DESCRIPTION
**Pull request by cdtj**
`w.trimLastPropertySeparator()` method was rewritten to remove all excessive commas and moved to `GetSerializedContent()` method to be called only once, on Entity serialization. 

Code:
```
requestBody.SetTeamsApp(app)
requestBody.SetId(app.GetId())
requestBody.SetAdditionalData(map[string]interface{}{
	"teamsApp@odata.bind": teamsAppPtr,
})
```

Before:
`{"id":"#ID","teamsApp@odata.bind":"#URL""teamsApp":{"id":"#ID","displayName":"#NAME","distributionMethod":"organization","externalId":"#EXTID"}}`
After:
`{"id":"#ID","teamsApp@odata.bind":"#URL","teamsApp":{"id":"#ID","displayName":"#NAME","distributionMethod":"organization","externalId":"#EXTID"}}`

### CODE REVIEW

**Review by baywet — CHANGES_REQUESTED** [review]
Thank you for the contribution. On top of the comments I added, would you mind adding a couple of unit tests for the scenarios around additional data please?
Would you mind adding [changelog](https://github.com/microsoft/kiota-serialization-json-go/blob/main/CHANGELOG.md) entries as well please?

**Comment by cdtj** [comment]
sounds fair, I'll re-review my code this weekend.

**Comment by cdtj** [comment]
I reviewed what I've tried to do.

1. AdditionalData now supports all types that previously were presented as pointers, 
also if an invalid data type is pushed now it throws an error, this might be good because I spent a few hours before I found out what really happening and why my data is just disappearing, 
at the same time, it might be harmful if the Entity is being generated by 3rd party and already contain some typed data that need to be forwarded next, this will throw an error and requires user to manually drop unsupported types, but I still don't think this is a real scenario.

2. Commas trimming, I tried different ways and found this one is the fastest:
```
cdtj@mbp kiota-serialization-json-go % go test -bench=. -count=5
goos: darwin
goarch: amd64
pkg: github.com/microsoft/kiota-serialization-json-go
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkMiltipleData-12                          702469              1693 ns/op
BenchmarkMiltipleData-12                          686631              1691 ns/op
BenchmarkMiltipleData-12                          667441              1689 ns/op
BenchmarkMiltipleData-12                          664492              1693 ns/op
BenchmarkMiltipleData-12                          648564              1698 ns/op
BenchmarkMiltipleDataWithoutTrimming-12           654273              1682 ns/op
BenchmarkMiltipleDataWithoutTrimming-12           633334              1689 ns/op
BenchmarkMiltipleDataWithoutTrimming-12           638966              1685 ns/op
BenchmarkMiltipleDataWithoutTrimming-12           662968              1694 ns/op
BenchmarkMiltipleDataWithoutTrimming-12           648363              1686 ns/op
PASS
ok      github.com/microsoft/kiota-serialization-json-go        12.639s
```
```
func BenchmarkMiltipleData(b *testing.B) {
	for i := 0; i < b.N; i++ {
		serializer := NewJsonSerializationWriter()
		value := "value"
		serializer.WriteStringValue("key", &value)
		pointer := "pointer"
		adlData := map[string]interface{}{
			"add1": "string",
			"add2": &pointer,
			"add3": []string{"foo", "bar"},
		}
		serializer.WriteAdditionalData(adlData)
		value2 := "value2"
		serializer.WriteStringValue("key2", &value2)
		serializer.GetSerializedContent()
	}
}
```
second benchmark is the same but `GetSerializedContent()` doesn't trim commas. 

This still might be wrong but the ways we used before need to keep in mind data that is already written and predict new data to put correct commas.

File: json_serialization_writer.go [hunk 1]
```diff
@@ -36,9 +37,19 @@ func (w *JsonSerializationWriter) writePropertySeparator() {
...
+		writerLen := len(w.writer)
+		if s == "," {
+			if idx <= writerLen-1 {
+				if idx == writerLen-1 {
+					w.writer[idx] = ""
+					continue
+				}
+				if w.writer[idx+1] == "]" || w.writer[idx+1] == "}" || w.writer[idx+1] == "," {
+					w.writer[idx] = ""
+				}
```
Comment (baywet) [inline json_serialization_writer.go:50]:
```suggestion
            if idx == writerLen-1 {
					w.writer[idx] = ""
		    } else if idx < writerLen-1 && (w.writer[idx+1] == "]" || w.writer[idx+1] == "}" || w.writer[idx+1] == ",") {
					w.writer[idx] = ""
```

File: json_serialization_writer.go [hunk 2]
```diff
@@ -36,9 +37,19 @@ func (w *JsonSerializationWriter) writePropertySeparator() {
...
+			}
+		}
```
Comment (baywet) [inline json_serialization_writer.go:52]:
```suggestion
		if s == "," && (idx == writerLen-1 || (idx < writerLen-1 && (w.writer[idx+1] == "]" || w.writer[idx+1] == "}" || w.writer[idx+1] == ",")) {
			w.writer[idx] = ""
		}
```

**Review by baywet — CHANGES_REQUESTED** [review]
thanks you for taking the time to make the changes and benchmark, one minor comment to simplify the code and reduce the number of branches.

**Comment by baywet** [comment]
also, for question 1, yes it's highly unlikely, and should that happen, the target type probably will implement IParsable (if generated from the API description)

**Comment by cdtj** [comment]
sorry if I clicked something wrong I'm new to github UI :)

suggestions commited

**Review by baywet — APPROVED** [review]

**Review by baywet — APPROVED** [review]
Thanks a lot for all the work!

**Pull request merged**